EncyclopediaTool

Aegonek opened the pull request:
Added EncyclopediaTool for tagging and untagging encyclopedia entries in text, heavily based on Vek TTT code.

**Notes:**

1. I moved entries data to encyclopedia.json file. I'm not sure how will it behave when published to NuGet. I think it should include the file as well, but it could be nice to test this via publishing pre-release version.

    This is mostly my personal preference to hold data in files instead of code; but knowing how to include resources in NuGet package may be helpful in the future.

2. I adjusted and added some encyclopedia entries.
3. I changed some rules of matching based on tests - it appears vanilla game only tags first occurence of encyclopedia entry in text, and I changed the code to work like that. I also refactored it a little to make it IMO easier to understand.

Linked to issue [Add tools for working with Encyclopedia tags.](https://github.com/WittleWolfie/WW-Blueprint-Core/issues/26) - to close after merging.

WittleWolfie commented on BlueprintCore/BlueprintCore.csproj:
```diff
@@ -94,6 +94,14 @@
+  
+  <ItemGroup>
+    <None Update="resources\encyclopedia.json">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
```
You might need to specify as content to work properly w/ NuGet. When you build the NuGet package is created--it's just a zip file so you can open and make sure the content is there.

Still you should verify this works w/ installing the NuGet package as well. It's unfortunately annoying but [here are the steps to install locally through VS](https://stackoverflow.com/questions/10240029/how-do-i-install-a-nuget-package-nupkg-file-locally).[]

Aegonek replied:
I resorted to adding content as EmbeddedResource to make it work. I tested that it works by creating a local project, adding a local nuget package and running a test.

I suppose it could be possible to make tests load WW-Blueprint-Core from local nuget package, but it sounds like the pain in the ass. I don't think it's necessary to do right now, do you?

WittleWolfie commented on BlueprintCore/BlueprintCore/Actions/Builder/ActionsBuilder.cs:
```diff
@@ -151,7 +151,7 @@ public ActionsBuilder Add(GameAction action)
     /// <remarks>It is recommended to only call this when adding an action type not supported by the builder.</remarks>
     /// 
     /// <param name="init">Optional initialization <see cref="Action"/> run on the action.</param>
-    public ActionsBuilder Add<A>(Action<A>? init = null) where A : GameAction, new()
```
Revert / sync to head where this is fixed. The generated code & template were out of sync. Same w/ ConditionsBuilder change.

Aegonek replied:
Fixed.

WittleWolfie commented on BlueprintCore/BlueprintCore/Internal/MatchContext.cs:
```diff
@@ -0,0 +1,32 @@
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+namespace BlueprintCore.BlueprintCore.Internal;
```
Remove redundant BlueprintCore from namespace.

Aegonek replied:
Fixed.

WittleWolfie commented on BlueprintCore/BlueprintCore/Internal/MatchExtensions.cs:
```diff
@@ -0,0 +1,14 @@
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+namespace BlueprintCore.BlueprintCore.Internal;
```
Remove redundant BlueprintCore from namespace.

Aegonek replied:
Fixed.

WittleWolfie commented on BlueprintCore/BlueprintCore/Utils/EncyclopediaTool.cs:
```diff
@@ -0,0 +1,115 @@
+using BlueprintCore.Utils;
+using BlueprintCore.BlueprintCore.Internal;
+
+namespace BlueprintCore.BlueprintCore.Utils;
```
Remove redundant BlueprintCore from namespace.

Aegonek replied:
Fixed.

WittleWolfie commented on BlueprintCore/BlueprintCore/Utils/EncyclopediaTool.cs:
```diff
@@ -0,0 +1,115 @@
+using BlueprintCore.BlueprintCore.Internal;
+
+namespace BlueprintCore.BlueprintCore.Utils;
+public static class EncyclopediaTool
```
Add a comment indicating this was modified from TTT & link to the repo.

Aegonek replied:
Done.

WittleWolfie commented:
Overall looks good and I appreciate the contribution!

Please update the instructions & documentation:

1. Additional steps needed, if any, to include the tagging JSON file in a mod release
2. Add example usage
3. Class / method comments on anything a mod dev would access

WittleWolfie commented:
I'm still getting back into the swing of this and dreading a little finishing the refactoring in progress :P Unfortunately the way I initially did code templating for the generator makes large scale improvements really painful so it needs to be done.

I'll take a more thorough look at this later this week, if everything is resolved by the weekend I'll likely push a new release with this.

Aegonek commented:
Oh yeah, I forgot to clean the imports :D Your way to test NuGet package is way better than mine. I'm gonna clean this up tomorrow.

Aegonek commented:
Actually I'm gonna do that later this week - I'm having a rather intense week right now 😆

Aegonek commented:
I've addressed all of your remarks. I'm not really satisfied with documentation, but I don't think adding long text sample to doc comments is such a good idea. A better way would probably be to add detailed example to doc website. I think it can go in next pull request though 😄I need to figure out how the docs work. They are generated using DocFX and hosted on GitHub pages, right? I've never used neither of those before.

Please see my comment about testing NuGet changes. If you have no objections to that, you can merge the pull request.
